### ISSUE TITLE
Argo: Thousands of games and database

### ISSUE THREAD

**Issue by davidjha** [c-to-the-fazzy/cardiathena#33]
# Description
Argo needs to be able to run thousands of games and interact with the database. This issue will look into getting that done. The method Save_game() will need to be modified from its current state to actually saving data to the database container on Argo. 

# Tasks

- [x] Game Interaction with database (Insert, retrieve) 
  - [x] Convert into MySQL format (Issue #32)
  - [x] Connect to SQL database
  - [x] Run games with slurm job scheduler
     - [x] Write slurm script to handle thousands of games
     - [x] Run slurm job

# Assignees 

@davidjha

**Comment by davidjha** [comment]
I'm having trouble connecting to the SQL container on Argo from a scheduled job or a job ran directly from a different compute node. I've tried to connect via socket located at /var/run/mysqld/mysqld.sock and /var/lib/mysql/mysql.sock. I was able to do this on my local machine, but I can't seem to do this anymore. Which makes me believe it has something to do with the container or the mysql configuration. I get this error when I run, [create_remote_user.sh](https://github.com/c-to-the-fazzy/cardiathena/blob/mysql-on-argo/argo/create_remote_admin_user.sh). :
<br></br>
`ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)`


I've tried changing permissions on the file and the directory via chmod 777, and adding this to the my.cnf file under [mysqld] and [mysql]:
<br></br>
`socket=/var/lib/mysql/mysql.sock`

My python connection config string includes: 
<br></br>
`unix_socket': '/path/to/mysqld.sock`

I've also tried connecting via hostname ie NODE0XX, My python connection config string would have this:
<br></br> 
`host': 'NODE0XX`
I get the error:
<br></br>
` Host xxx.xx.xxx.xxx is not allowed to connect to this MySQL server.`
I found this: [](https://stackoverflow.com/questions/1559955/host-xxx-xx-xxx-xxx-is-not-allowed-to-connect-to-this-mysql-server) Which I am going to try, although the create_remote_user script does include the @'%' in 'remote_usr'@'%' to allow connections from anywhere.

**Comment by c-to-the-fazzy** [comment]
Hi David, I'm not an expert on this kind of thing, but I believe `mysqld.sock` is not going to be available when the client container/job is on a different node.  

I think the hostname connection is what you're looking for.  I'm not sure why that connection would be refused though - are you setting a username/password anywhere?

**Comment by davidjha** [comment]
Ok, I thought it would work from the different nodes because they share or can access the same filesystem.

I am setting root@localhost (default settings) and a remote_usr with admin privliedges (for testing). The remote_usr is created :
`"GRANT ALL PRIVILEGES ON *.* TO 'remote_usr'@'%' IDENTIFIED BY '$RANDOM_PASS' WITH GRANT OPTION"`

Also when I was running the mysql container on my  local machine, I ran a test looping through the game 30000 times, the insertions hit a hard bottleneck somewhere? It slows down by a lot, but I was able to create about 2GB of data after several hours + multiple rebuilds/restarts on the container to try to make it run faster.

**Comment by c-to-the-fazzy** [comment]
OK, I don't have any additional advice at this time regarding the username/password issue.  

But, the data generation rates seem promising to start!  Several GB corresponding to 10's of thousands of games over the course of several hours is a great start.  It'll be important in the long run to understand the cause of the bottleneck you observed, but for now I wouldn't sweat it too much.

**Comment by davidjha** [comment]
Ok thanks Carlo,
I figured out the the username/password problem, I executed the create_remote_usr script before the server in the container is started. Therefore, there is no running server to connect to create a user that can connect remotely. I was also getting the cannot connect to socket error:
`ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock`

The ` Host xxx.xx.xxx.xxx is not allowed to connect to this MySQL server.` error is solved once the remote_usr is created and the password set, I am able to connect to the database node from any of the other compute nodes. I had to set the connection to the ip or hostname of the database node.

I ran the game 30000 times, it took about 5 hours running 16 instances at a time. Checking the database, I had all the entries by querying an id column.

**Comment by irobert4** [comment]
So does each instance run the game 30000 times, or is it 30000 games spread across 16 instances?

Also, how much data are you getting now?

**Comment by davidjha** [comment]
Its 30000 games ran at 16 at a time spread across random nodes. The file is 4.7GB with 34000 entries. The 4000 is from previous test runs.

**Comment by irobert4** [comment]
Ok, did all of these instances start fast and then slow down to a crawl again?

**Comment by davidjha** [comment]
I can't really tell, once the job is submitted to the scheduler I can see the output files produced. Each game produces a separate output file. So, I'm getting 30000 output files which I can't really tell what's going on because the filesystem slows everything down when trying to do anything in that directory. So if I were to ls or list the files in the directory, it takes a long time. 

There's a command sacct -X which lists all pending/completed jobs, but I think the slurm database that keeps track of this is bogged down also. I get an error saying I can't access the slurmdb.

However, this runs a lot faster then my local machine.

**Comment by aphelps4** [comment]
Each file should have timestamps when created, right? If so, you could run it with only one game going and compare the time difference between the first two files and the final two files to check if there is any slowdown going on.

**Comment by davidjha** [comment]
I don't think running one game would show anything meaningful right now. The code saves all of the states from one game as one entry/row in the database.

However I ran the game 1000 times at 16 at a time and produced this result:

[https://raw.githubusercontent.com/c-to-the-fazzy/cardiathena/mysql-on-argo/argo/runtimes](https://raw.githubusercontent.com/c-to-the-fazzy/cardiathena/mysql-on-argo/argo/runtimes)

So going off of what you said, I took the difference of timestamps of the first and last 2 files and I got this:

First output file:                    16:43:48.388590079
Second output file:               16:43:49.995619216

Difference: 1.607029137

Third:                                    16:43:50.192622788
Fourth:                                  16:43:50.193622808 

Difference: 0.00100002

999th output file:                    17:15:22.230862153
1000th output file:                  17:15:22.265862785

Difference: 0.035000632

The difference between the first and second output files is a bit of an anomaly.
